File: samples/JavascriptSamples/assets/js/samples/file_sample.js
import { Generator } from '../lib/salmon-core/salmon/generator.js';
import { Integrity } from '../lib/salmon-core/salmon/integrity/integrity.js';
import { AesFile } from '../lib/salmon-fs/salmonfs/file/aes_file.js';

export class FileSample {
	static BUFFER_SIZE = 256 * 1024; // recommended buffer size aligned to internal buffers
	
    static async encryptTextToFile(text, key, integrityKey, file) {
        // encrypt to a file, the AesFile has a virtual file system API
        print( "Encrypting text to file: " + file.getName());

        let data = new TextEncoder().encode(text);
		
        // Always request a new random secure nonce
        let nonce = Generator.getSecureRandomBytes(8); // 64 bit nonce

        let encFile = new AesFile(file);
        encFile.setEncryptionKey(key);
        encFile.setRequestedNonce(nonce);

		if(integrityKey)
			encFile.setApplyIntegrity(true, integrityKey, Integrity.DEFAULT_CHUNK_SIZE);
		else
			encFile.setApplyIntegrity(false);
		
        let encStream = await encFile.getOutputStream();

		// now write the data you want to decrypt
		// it is recommended to use a large enough buffer while writing the data
		// for better performance
		let totalBytesWritten = 0;
        while(totalBytesWritten < data.length) {
			let length = Math.min(data.length - totalBytesWritten, FileSample.BUFFER_SIZE);
			await encStream.write(data, totalBytesWritten, length);
			totalBytesWritten += length;
		}
        await encStream.flush();
        await encStream.close();
    }
	
	static async decryptTextFromFile(key, integrityKey, file) {
		print( "Decrypting text from file: " + file.getName());
		
        // Wrap the file with a AesFile
		// the nonce is already embedded in the header
        let encFile = new AesFile(file);

		// set the key
        encFile.setEncryptionKey(key);

		if(integrityKey)
			encFile.setVerifyIntegrity(true, integrityKey);
		else
			encFile.setVerifyIntegrity(false);
		
		// open a read stream
        let decStream = await encFile.getInputStream();

        // decrypt the data
		let decData = new Uint8Array(await decStream.getLength());
		let totalBytesRead = 0;
		let bytesRead = 0;
        while((bytesRead = await decStream.read(decData, totalBytesRead, FileSample.BUFFER_SIZE)) > 0) {
			totalBytesRead += bytesRead;
		}
		
        let decText = new TextDecoder().decode(decData.slice(0, totalBytesRead));
        await decStream.close();
		
		return decText;
    }
}

File: samples/PythonSamples/samples/file_sample.py
#!/usr/bin/env python3
from salmon_core.salmon.integrity.integrity import Integrity
from salmon_core.salmon.generator import Generator
from salmon_fs.fs.file.ifile import IFile
from salmon_fs.salmonfs.file.aes_file import AesFile


class FileSample:
    BUFFER_SIZE = 256 * 1024  # recommended buffer size aligned to internal buffers

    @staticmethod
    def encrypt_text_to_file(text: str, key: bytearray, integrity_key: bytearray, file: IFile):
        # encrypt to a file, the AesFile has a virtual file system API
        print("Encrypting text to file: " + file.get_name())

        data = bytearray(text.encode())

        # Always request a new random secure nonce
        nonce = Generator.get_secure_random_bytes(8)  # 64 bit nonce

        enc_file = AesFile(file)
        enc_file.set_encryption_key(key)
        enc_file.set_requested_nonce(nonce)

        if integrity_key:
            enc_file.set_apply_integrity(True, integrity_key, Integrity.DEFAULT_CHUNK_SIZE)
        else:
            enc_file.set_apply_integrity(False)

        enc_stream = enc_file.get_output_stream()

        # now write the data you want to decrypt
        # it is recommended to use a large enough buffer while writing the data
        # for better performance
        total_bytes_written = 0

        while total_bytes_written < len(data):
            length = min(len(data) - total_bytes_written, FileSample.BUFFER_SIZE)
            enc_stream.write(data, total_bytes_written, length)
            total_bytes_written += length
        enc_stream.flush()
        enc_stream.close()

    @staticmethod
    def decrypt_text_from_file(key: bytearray, integrity_key: bytearray | None, file: IFile) -> str:
        print("Decrypting text from file: " + file.get_name())

        # Wrap the file with a AesFile
        # the nonce is already embedded in the header
        enc_file = AesFile(file)

        # set the key
        enc_file.set_encryption_key(key)

        if integrity_key:
            enc_file.set_verify_integrity(True, integrity_key)
        else:
            enc_file.set_verify_integrity(False, None)

        # open a read stream
        dec_stream = enc_file.get_input_stream()

        # decrypt the data
        dec_data = bytearray(dec_stream.get_length())
        total_bytes_read = 0
        while bytes_read := dec_stream.read(dec_data, total_bytes_read, FileSample.BUFFER_SIZE) > 0:
            total_bytes_read += bytes_read

        dec_text = dec_data[0:total_bytes_read].decode()
        dec_stream.close()

        return dec_text
